File: built-in plugins dsp/mb-bell-churchbell.cpp
/**
 * MB Church Bell
 * Category : instrument
 * Type     : bell
 * Version  : 1.0.0
 * Author   : Max Booster
 * Desc     : Deep resonant church bell with complex partials
 *
 * Part of Max Booster Built-In Plugins DSP
 */

#ifndef MB_BELL_CHURCHBELL_H
#define MB_BELL_CHURCHBELL_H

#include <algorithm>
#include <cmath>
#include <cstring>
#include "PluginBase.h"

class MbBellChurchbell : public PluginBase {
public:
    static constexpr const char* PLUGIN_ID      = "mb-bell-churchbell";
    static constexpr const char* PLUGIN_NAME    = "MB Church Bell";
    static constexpr const char* PLUGIN_TYPE    = "bell";
    static constexpr const char* PLUGIN_CATEGORY = "instrument";
    static constexpr const char* VERSION         = "1.0.0";

    struct Parameters {
    float size = 0.6f;  // range [0, 1]
    float decay_time = 0.8f;  // range [0, 1]
    float partials = 0.5f;  // range [0, 1]
    float volume = 0.8f;  // range [0, 1]
    };

    MbBellChurchbell() = default;
    ~MbBellChurchbell() override = default;

    void setSampleRate(double sampleRate) override {
        sampleRate_ = sampleRate;
        reset();
    }

    void reset() override {
        std::memset(buffer_, 0, sizeof(buffer_));
    }

    void process(float* left, float* right, int numSamples, Parameters params) {
        params.size = std::clamp(params.size, 0f, 1f);
        params.decay_time = std::clamp(params.decay_time, 0f, 1f);
        params.partials = std::clamp(params.partials, 0f, 1f);
        params.volume = std::clamp(params.volume, 0f, 1f);
        for (int i = 0; i < numSamples; ++i) {
            left[i]  = processSample(left[i],  params);
            right[i] = processSample(right[i], params);
        }
    }

private:
    double sampleRate_ = 44100.0;
    float  buffer_[65536] = {};

    inline float processSample(float input, const Parameters& params) {
        // DSP implementation for MB Church Bell
        return input;
    }
};

#endif // MB_BELL_CHURCHBELL_H

File: built-in plugins dsp/mb-bell-churchbell.ts
import type { PluginDefinition } from "../server/services/pluginHostService";

const MbBellChurchbellPlugin: PluginDefinition = {
  id: "mb-bell-churchbell",
  slug: "mb-bell-churchbell",
  name: "MB Church Bell",
  category: "instrument",
  type: "bell" as any,
  version: "1.0.0",
  description: "Deep resonant church bell with complex partials",
  author: "Max Booster",
  grade: "A",
  oscillators: [
    { type: "sine", detune: 0, gain: 0.35 },
    { type: "sine", detune: 1200, gain: 0.25 },
    { type: "sine", detune: 1550, gain: 0.2 },
    { type: "sine", detune: 2400, gain: 0.2 },
  ],
  envelope: { attack: 0.005, decay: 4.0, sustain: 0.1, release: 3.0 },
  parameters: [
    {
      id: "size",
      name: "Bell Size",
      type: "float",
      defaultValue: 0.6,
      minValue: 0,
      maxValue: 1,
      automatable: true,
    },
    {
      id: "decay_time",
      name: "Decay",
      type: "float",
      defaultValue: 0.8,
      minValue: 0,
      maxValue: 1,
      automatable: true,
    },
    {
      id: "partials",
      name: "Partials",
      type: "float",
      defaultValue: 0.5,
      minValue: 0,
      maxValue: 1,
      automatable: true,
    },
    {
      id: "volume",
      name: "Volume",
      type: "float",
      defaultValue: 0.8,
      minValue: 0,
      maxValue: 1,
      automatable: true,
    },
  ],
  defaultPreset: { size: 0.6, decay_time: 0.8, partials: 0.5, volume: 0.8 },
};

export default MbBellChurchbellPlugin;
